### PR TITLE
Added prompting along with e5 instruct

### PR DESCRIPTION
**Pull request by KennethEnevoldsen**
All encode calls now include a prompt_name (following the variable name in sentence transformer's encode), which provided the task name. The model can then use a custom prompt pr. task if they wish (or encode based on e.g. task type).

Additionally to test it out I have also added the two e5 instruct models.

## Checklist

- [x] Run tests locally to make sure nothing is broken using `make test`. 
- [x] Run the formatter to format the code using `make lint`. 

### Adding a model checklist

 - [x] I have filled out the ModelMeta object to the extent possible
 - [x] I have ensured that my model can be loaded using
   - [x] `mteb.get_model(model_name, revision_id)` and
   - [x] `mteb.get_model_meta(model_name, revision_id)`
 - [x] I have tested the implementation works on a representative set of tasks.

### CODE REVIEW

File: mteb/models/instructions.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+# scandinavian embedding benchmark: https://github.com/KennethEnevoldsen/scandinavian-embedding-benchmark/blob/c8376f967d1294419be1d3eb41217d04cd3a65d3/src/seb/registered_models/e5_instruct_models.py
+# e5 documentation: https://github.com/microsoft/unilm/blob/9c0f1ff7ca53431fe47d2637dfe253643d94185b/e5/utils.py#L106
+DEFAULT_PROMPTS = {
+    "STS": "Retrieve semantically similar text",
+    "Summarization": "Given a news summary, retrieve other semantically similar summaries",
+    "BitextMining": "Retrieve parallel sentences.",
+    "Classification": "Classify user passages",
+    "Clustering": "Identify categories in user passages",
+    "Reranking": "Retrieve text based on user query.",
+    "Retrieval": "Retrieve text based on user query.",
```
Comment (imenelydiaker) [inline mteb/models/instructions.py:15]: We're using the same prompts for retrieval and reranking?

File: mteb/models/instructions.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
+"""This specifies the default instruction for tasks within MTEB. These are optional to use and some models might want to use their own instructions."""
+
+import mteb
+
+# prompt are derived from:
+# scandinavian embedding benchmark: https://github.com/KennethEnevoldsen/scandinavian-embedding-benchmark/blob/c8376f967d1294419be1d3eb41217d04cd3a65d3/src/seb/registered_models/e5_instruct_models.py
+# e5 documentation: https://github.com/microsoft/unilm/blob/9c0f1ff7ca53431fe47d2637dfe253643d94185b/e5/utils.py#L106
+DEFAULT_PROMPTS = {
+    "STS": "Retrieve semantically similar text",
```
Comment (imenelydiaker) [inline mteb/models/instructions.py:9]: Missing one task: "PairClassification"

File: mteb/evaluation/evaluators/BitextMiningEvaluator.py [hunk 3]
```diff
@@ -84,12 +87,12 @@ def _compute_metrics(
...
             "precision": precision_score(
-                labels, predictions, zero_division=0.0, average="weighted"
+                labels, predictions, zero_division=0, average="weighted"
             ),
             "recall": recall_score(
-                labels, predictions, zero_division=0.0, average="weighted"
+                labels, predictions, zero_division=0, average="weighted"
             ),
-            "f1": f1_score(labels, predictions, zero_division=0.0, average="weighted"),
+            "f1": f1_score(labels, predictions, zero_division=0, average="weighted"),
```
Comment (imenelydiaker) [inline mteb/evaluation/evaluators/BitextMiningEvaluator.py:95]: In classification we have a lot of class imbalance, should we also use weighted metrics?

File: mteb/evaluation/evaluators/normalize_encode.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
+from __future__ import annotations
```
Comment (imenelydiaker) [inline mteb/evaluation/evaluators/normalize_encode.py:1]: I find the name of this module a bit confusing: the code doesn't really normalize embeddings right? It just converts them to an array. Maybe rename the module to `model_encode` just as the function or `encode`

Comment (imenelydiaker) [reply mteb/evaluation/evaluators/normalize_encode.py:1]: Wouldn't it be better to add the function to `Evaluator` class and inherit it in other child evaluators?

**Review by imenelydiaker — COMMENTED** [review]
Looks great! Left some comments 🙂

File: mteb/evaluation/evaluators/BitextMiningEvaluator.py (hunk 3 above)
Comment (KennethEnevoldsen) [reply mteb/evaluation/evaluators/BitextMiningEvaluator.py:95]: We can add it but I wouldn't change it.

Comment (KennethEnevoldsen) [reply mteb/evaluation/evaluators/BitextMiningEvaluator.py:95]: Added ap_weighted and f1_weighted

File: mteb/models/instructions.py (hunk 2 above)
Comment (KennethEnevoldsen) [reply mteb/models/instructions.py:9]: added

File: mteb/models/e5_instruct.py [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+        prompt_name: str | None = None,
+        batch_size: int = 32,
+        encode_type: EncodeTypes = "query",
+        **kwargs: Any,  # noqa
+    ) -> np.ndarray:
+        if self.max_batch_size and batch_size > self.max_batch_size:
+            batch_size = self.max_batch_size
+        batched_embeddings = []
+        if prompt_name is not None:
+            instruction = task_to_instruction(prompt_name)
```
Comment (Muennighoff) [inline mteb/models/e5_instruct.py:100]: Nice, I think I'll also add this to GritLM after this PR though I'd have to subclass it I think and turn the `prompt_name` kwarg into the `instruction` kwarg used for GritLM (https://github.com/ContextualAI/gritlm/blob/da37ccbace1aa4f4bf26273e4e9a7cea705ae951/gritlm/gritlm.py#L98)

**Review by Muennighoff — COMMENTED** [review]
Amazing 🙌

File: mteb/evaluation/evaluators/model_encode.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+def model_encode(
+    sentences: Sequence[str], *, model: Encoder, task_name: str, **kwargs
+) -> np.ndarray:
+    kwargs["prompt_name"] = task_name
+    if hasattr(model, "prompts") and task_name not in model.prompts:  # type: ignore
+        logger.info(
+            f"Prompt {task_name} not found in model prompts. Removing prompt_name argument."
+        )
+        kwargs.pop("prompt_name")
```
[the remark this answers is not in the document]
Comment (KennethEnevoldsen) [reply mteb/evaluation/evaluators/model_encode.py:22]: We assume that the model can take the `prompt_name` argument (or at least kwargs), if not they will get an error.

This is simply to handle sentence transformers models which will throw an error if they do not handle the prompt. A solution might be to simply "wrap" the model's encode fn if it is a sentence transformer.

File: mteb/models/instructions.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (KennethEnevoldsen) [reply mteb/models/instructions.py:15]: Yea probably - same as retrieval?

[the remark this answers is not in the document]
Comment (Muennighoff) [reply mteb/models/instructions.py:15]: Yes I think that would make sense

File: mteb/models/__init__.py [hunk 7]
```diff
@@ -98,7 +98,7 @@ def model_meta_from_sentence_transformers(model: SentenceTransformer) -> ModelMe
     return meta
 
 
-model_modules = [e5_models, sentence_transformers_models]
+model_modules = [e5_models, sentence_transformers_models, e5_instruct]
```
Comment (Muennighoff) [inline mteb/models/__init__.py:101]:
```suggestion
model_modules = [e5_instruct, e5_models, sentence_transformers_models]
```

File: mteb/models/instructions.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Muennighoff) [reply mteb/models/instructions.py:15]: Actually @orionw do you mind having a quick look at this PR & whether or not anything should change w.r.t. Instruction Retrieval?

File: mteb/evaluation/evaluators/ClassificationEvaluator.py [hunk 8]
```diff
@@ -150,10 +162,11 @@ def __init__(
         y_train,
         sentences_test,
         y_test,
-        k=1,
-        batch_size=32,
-        limit=None,
-        **kwargs,
+        task_name: str,
```
Comment (Muennighoff) [inline mteb/evaluation/evaluators/ClassificationEvaluator.py:165]: Do we need to make `task_name` mandatory everywhere? Can we not have it default to `None`? Some people are importing these evaluators from MTEB & we will break their code.

File: tests/test_RetrievalEvaluator.py [hunk 9]
```diff
@@ -13,7 +13,7 @@ def setup_method(self):
 
         setup_method is invoked for every test method of a class.
         """
-        self.evaluator = RetrievalEvaluator()
+        self.evaluator = RetrievalEvaluator(task_name="test")
```
Comment (Muennighoff) [inline tests/test_RetrievalEvaluator.py:16]: Like earlier comment, I think `task_name` should not be mandatory if there are cases where it is not needed like here.

Also, maybe we should have a test to ensure that prompting works? E.g. test it produces different results on a toy model with the latest sentence transformers or sth?

**Review by Muennighoff — COMMENTED** [review]
Sorry for the delay! It looks great tho would love if @orionw could have a quick look whether or not this aligns with Retrieval w/ Instructions 🙌

**Comment by orionw** [comment]
Looks great @KennethEnevoldsen! Using the same as Retrieval for InstructionRetrieval is great and it will be nice for all the many models using it these days.  For many tasks in InstructionRetrieval the models did better with the prefix and the instruction so I added them both manually, glad to have it be part of this.

File: mteb/models/instructions.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (KennethEnevoldsen) [reply mteb/models/instructions.py:15]: I have added the instruction - we can always change it later

File: mteb/evaluation/evaluators/ClassificationEvaluator.py (hunk 8 above)
Comment (KennethEnevoldsen) [reply mteb/evaluation/evaluators/ClassificationEvaluator.py:165]: I have change it to default to None

File: mteb/models/instructions.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (KennethEnevoldsen) [reply mteb/models/instructions.py:15]: I have added the retrieval instructions to this task

File: mteb/evaluation/evaluators/ClassificationEvaluator.py (hunk 8 above)
Comment (KennethEnevoldsen) [reply mteb/evaluation/evaluators/ClassificationEvaluator.py:165]: I have change it to default to None

File: mteb/encoder_interface.py [hunk 10]
```diff
@@ -1,23 +1,26 @@
...
+
 
 @runtime_checkable
 class Encoder(Protocol):
-    """The interface for an encoder in MTEB."""
+    """The interface for an encoder in MTEB. In general we try to keep this interface aligned with sentence-transformers."""
 
     def encode(
-        self, sentences: list[str], prompt: str, **kwargs: Any
+        self, sentences: Sequence[str], *, prompt_name: str | None = None, **kwargs: Any
```
[the remark this answers is not in the document]
Comment (KennethEnevoldsen) [reply mteb/encoder_interface.py:16]: edit: it seems like it uses `instruction`

File: tests/test_RetrievalEvaluator.py (hunk 9 above)
Comment (KennethEnevoldsen) [reply tests/test_RetrievalEvaluator.py]: Completely agree. I have added a test to ensure that the task_name is corrected and passed to models where it is relevant and does not influence the sentence transformer, which does not take the prompt.

**Pull request merged**